flow/FileTraceLogWriter.cpp

```diff
@@ -79,6 +82,8 @@ void FileTraceLogWriter::write(const std::string& str) {
 			remaining -= ret;
 			ptr += ret;
 		} else {
+			issues->addAndExpire("trace_log_file_write_error");
+			fprintf(stderr, "Unexpected error [%d] when flushing trace log.\n", errno);
```

Comment:
It looks like this and the failure to open trace files both get retried indefinitely, which probably mean that they will also trigger the write thread liveness check. This is mainly just an observation, it's probably fine to log both issues as that can help pinpoint the error better.

Reply:
The same comment applies about whether the expiration is sufficient to avoid this disappearing before it gets resolved. I think the duration of the sleep means that the only way for that to happen is if `__write` blocks, which should mean that at least the liveness check issue persists, but we'd lose this extra detail.

Reply:
> which probably mean that they will also trigger the write thread liveness check

Can you elaborate on this? Currently write thread liveness check is done by posting a Ping action to the thread and wait for the signal to be set.

>The same comment applies about whether the expiration is sufficient to avoid this disappearing before it gets resolved. I think the duration of the sleep means that the only way for that to happen is if __write blocks, which should mean that at least the liveness check issue persists, but we'd lose this extra detail.

Yeah this is true. One possible solution is that for all the expired issues, instead of removing them when being retrieved, we can still include them into the returned issues list but mark them as `expired` as an indicator of what happened recently, after that "first" retried, we then remove them.

Reply:
@ajbeamon

Reply:
> Can you elaborate on this?

If I understand correctly, in both of these error cases the writer thread will block until the problem resolves itself, in which case the liveness check will trigger because the ping can't occur.

> One possible solution is that for all the expired issues, instead of removing them when being retrieved, we can still include them into the returned issues list but mark them as `expired` as an indicator of what happened recently, after that "first" retrieval, we then remove them.

Another possibility is to keep state about whether we're in the error state or not, adding the issue if we are in the error state. We could enter the state when we encounter the error and we leave it when the problem resolves.

Reply:
> If I understand correctly, in both of these error cases the writer thread will block until the problem resolves itself, in which case the liveness check will trigger because the ping can't occur.

Oh right. In that case, without any `error state` or something like that, the `trace_log_writer_thread_unresponsive` would override previous errors like failure to flush file.

> Another possibility is to keep state about whether we're in the error state or not, adding the issue if we are in the error state. We could enter the state when we encounter the error and we leave it when the problem resolves.

I like this idea. Let me see if I can quickly add it.

Reply:
I ended up changing the issue resolve to be a manual behavior performed at the caller site for the simplicity. If this turns out to be adding too much hassle, we can come back and change it later to something more sophisticated.